ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/StructSerializationModule.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.rest.util;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import java.io.IOException;
+import java.util.Collections;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.json.JsonConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StructSerializationModule extends SimpleModule {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StructSerializationModule.class);
+  private static final JsonConverter jsonConverter = new JsonConverter();
+
+  public StructSerializationModule() {
+    super();
+    jsonConverter.configure(Collections.singletonMap("schemas.enable", false), false);
+    addSerializer(Struct.class, new StructSerializationModule.Serializer());
+  }
+
+  static class Serializer extends JsonSerializer<Struct> {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Override
+    public void serialize(
+        final Struct struct,
+        final JsonGenerator jsonGenerator,
+        final SerializerProvider serializerProvider
+    ) throws IOException {
+      struct.validate();
+      jsonGenerator.writeObject(objectMapper.readTree(
```

Comment:
Am I right in thinking `jsonConverter.fromConnectData()` converts the `struct` to a `Struct` to a `String`, then `objectMapper.readTree()` converts the `String` into a `JsonNode`, then `jsonGenerator.writeObject()` converts the node into a String....?

So thats `Struct`->`String`->`JsonNode`->`String` - sounds like madness to me! ;)

Try this instead:

```
static class StructSerializer extends JsonSerializer<Struct> {

    @Override
    public void serialize(final Struct struct,
                          final JsonGenerator gen,
                          final SerializerProvider serializers) throws IOException {
      struct.validate();

      gen.writeStartObject();

      final Schema schema = struct.schema();
      for (final Field field : schema.fields()) {
        gen.writeObjectField(field.name(), struct.get(field));
      }

      gen.writeEndObject();
    }
  }
```

:D

Reply:
I wanted to make sure the generated json is consistent with Connect json so I used connect jsonConverter. Another reason to use it is that it handles nested structs too. JsonGenerator won't serialize structs with struct fields.

Reply:
JsonConverter is nice because it takes care of serializing connect data types that aren't readily json-serialized, and is also what we use to write records to json-formatted output topics. Admittedly we don't currently use any types in ksql that are hard to serialize, but that could easily change - for example, we could support maps with non-string keys.

Reply:
The very fact that you're writing a serializer, to customize an `ObjectMapper`, that requires an `ObjectMapper` internally, should be firing alarm bells.

Your current code serializes, deserializes and then re-serializes the object again  - it's madness!

The code I give above also handles nested types.  But does so using the normal patterns for customizing an `ObjectMapper`.  

Our rest app and other parts of our system use the Jackson ObjectMapper, not Connect's JsonConverter.  The use of ObjectMapper gives us the ability to handle Connect and non-Connect types.  I don't think it's possible to convert everything over to use JsonConverter as we have non-Connect types too.  The hybrid code you have here is ... weird! lol. And really inefficient.

We already have `SchemaMapper`. It feels to me that the StructSerializer class, as I have above, should be added to `SchemaMapper`, rather than a different Struct specific Module.  

Then, if you think there is a strong case to use the Connect JsonCoverter thing, then lets discuss and potentially put that is as a separate PR.

Reply:
Struct isn't a serializable type itself. Connect Converter is an API purpose-built for serializing Struct. It takes care of a few cases that the vanilla ObjectMapper doesn't gracefully handle. Two cases I know of are maps with non-string keys, and logical types. To be  fair we don't actually support these from KSQL at the moment. But the underlying point is why reimplement something that we already have an API for?

I totally agree that the mechanics are weird. But I don't see another way to get jackson to use the converters to serialize structs. The converter api goes from struct to binary. It doesn't return an intermediate object like JsonNode.

I'm not as concerned about the extra overhead as this path isn't performance sensitive at the moment. It's only used for streaming back rows via the rest api. If that becomes a real way for people to get data out of ksql from their apps, or we expect to have lots of these going on all the time the conversation changes of course, but for now I don't think thats the case.

> We already have SchemaMapper. It feels to me that the StructSerializer class, as I have above, should be added to SchemaMapper, rather than a different Struct specific Module.

SchemaMapper has been removed since it wasn't actually being used by anything other than the websocket endpoint, which was changed to use the ksql schema description entities to be consistent with the rest of the api.